src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,206 @@
+/// check if there are previous method calls to determine if an arm is borrowing or taking ownership
+/// of the returned type. If there are subsequent method calls, we can infer that the lifetime of
+/// the returned type is *potentially* surprising, so return true. In all other cases, return false.
+fn is_expr_with_significant_drop<'a, 'tcx>(
+    expression: &Expr<'a>,
+    context: &LateContext<'tcx>,
+    is_last_method_call: bool,
+) -> bool {
+    if is_last_method_call || !is_expr_with_significant_drop_attribute(expression, context) {
+        match expression.peel_drop_temps().kind {
+            ExprKind::MethodCall(_path, exprs, _source) => is_expr_with_significant_drop(&exprs[0], context, false),
+            _ => false,
+        }
+    } else {
+        true
+    }
+}
+
+/// Check if an expression has the `has_significant_drop` attribute or if any of its subtypes recursively
+/// have the attribute. If found, this indicates the expression may have a surprising lifetime.
+fn is_expr_with_significant_drop_attribute<'a, 'tcx>(expression: &Expr<'a>, context: &LateContext<'tcx>) -> bool {
+    let self_ty = context.typeck_results().expr_ty(expression);
+
+    if let Some(def_id) = get_def_id(self_ty) {
+        if contains_significant_drop_attr(context.tcx.get_attrs(def_id)) {
+            return true;
+        }
+    }
+
+    has_subtype_with_significant_drop(self_ty, &context)
+}
+
+/// Check if any of the `Normal` attributes match "has_significant_drop"
+fn contains_significant_drop_attr(attributes: Attributes<'_>) -> bool {
+    for a in attributes {
+        match &a.kind {
+            AttrKind::Normal(ai, _) => {
+                for s in &ai.path.segments {
+                    if s.ident.name.as_str() == "has_significant_drop" {
+                        return true;
+                    }
+                }
+            }
+            _ => {}
+        };
+    }
+    false
+}
+
+/// In addition to checking the return type itself, we want to check the return types fields and
+/// generic arguments. If a subfield or a generic argument has a significant drop, it will could
+/// result in surprising behavior for the caller due to the lifetime of the temporary being extended.
+fn has_subtype_with_significant_drop<'tcx>(expr_type: Ty<'tcx>, context: &LateContext<'tcx>) -> bool {
+    match expr_type.kind() {
+        rustc_middle::ty::Adt(a, b) => {
+            for f in a.all_fields() {
+                let field_type = f.ty(context.tcx, b);
+                if let Some(def_id) = get_def_id(expr_type) {
+                    if contains_significant_drop_attr(context.tcx.get_attrs(def_id))
+                        || has_subtype_with_significant_drop(field_type, context)
+                    {
+                        return true;
+                    }
+                }
+            }
+
+            if let Some(type_list) = b.try_as_type_list() {
```

Comment:
This doesn't work if the subst list has a `const` or `lifetime` in between generic types. IIUC the `try_as_type_list` function returns `None` in that case. But there could still be types with significant drops in the list of generics.

Reply:
I did not realize that! Thank you! Updated to use `unpack` which I think works here.